microSWIFT/config.txt
Configuration file for the microSWIFT wave bouy.

-----------------------------
Duty Cycle Length
-----------------------------
The length of time of the total duty cycle. The duty cycle is defined 
as the total time for both the record and send windows for the 
microSWIFT. The duty cycle is defined in units of MINUTES. It must be 
an integer value, less than 60 minutes, and a factor of 60.
Default value is 60 minutes.

duty_cycle_length = 60

-----------------------------
Record Window Length
-----------------------------
The length of time that the instruments will be recording. This time 
MUST be shorter than the defined length of the duty cycle. The record
length is defined in units of MINUTES and it must be an integer value.
Default value is 50 minutes.

record_window_length = 50

-----------------------------
GPS Sampling Frequency
-----------------------------
The sampling frequency of the GPS. This is the rate at which the GPS 
will sample during the record window. The gps sampling frequency is
defined in units of HZ (samples/second). This value can range from 
1 to 4 Hz. This value must also be an integer. Default value is 4 Hz.

gps_sampling_frequency = 4

-----------------------------
IMU Sampling Frequency
-----------------------------
The sampling frequnecy of the IMU. This is the rate at which the IMU 
will sample during the record window. The imu sampling frequency is 
defined in units of HZ (samples/second). This value can range from
1 to 30 Hz. This value must also be an integer. Default value is 12 Hz.

imu_sampling_frequency = 12

utils/IMU settings.txt
FXOS8700C Accelerometer/Magnetometer

CTL_REG1 (address 0x2A): write 0010 0101 = 0x25 to control register 1
	[7-6]: aslp_rate=00: auto-wake sample frequency when device in sleep mode (sleep mode not used)
	[5-3]: dr=100: output data rate 25Hz (for hybrid mode)
	[2]: lnoise=1: reduced noise mode active (2g and 4g range only)
	[1]: f_read=0: normal mode.  when f_read=1, only 8-bit MSB registers are read
	[0]: active=1: active mode

CTL_REG2 (address 0x2B): write 0000 0010 = 0x02 to control register 2
	[7]: st=0: self test function disable
	[6]: rst=0: software reset function disable
	[5]: -
	[4-3]: smods=00: accelerometer sleep mode oversample ratio (OSR) mode selection (sleep mode not used)
	[2]: slpe=0: auto sleep mode disabled
	[1-0]: mods=10: accelerometer wake mode OSR mode selection, high resolution mode

XYZ_DATA_CFG (address 0x0E): write 0000 0001 = 0x01 to XYZ_DATA_CFG register
	[7]: reserved
	[6]: reserved
	[5]: reserved
	[4]: hpf_out=0
	[3]: reserved
	[2]: reserved
	[1-0]: fs=00 for accelerometer range of +/-2g range with 0.244mg/LSB

M_CTRL_REG1 (address 0x5B): write 0001 1111 = 0x1F to magnetometer control register 1
	[7]: m_acal=0: auto calibration disabled
	[6]: m_rst=0: no one-shot magnetic reset
	[5]: m_ost=0: no one-shot magnetic measurement
	[4-2]: m_os=111=7: 8x oversampling (for 200Hz) to reduce magnetometer noise
	[1-0]: m_hms=11=3: select hybrid mode with accel and magnatometer active

M_CTRL_REG2 (address 0x5C): write 0010 0000 = 0x20 to magnetometer control register 2
	[7]: reserved
	[6]: reserved
	[5]: hyb_autoinc_mode=1 to map the magnetometer registers to follow the accelerometer registers
	[4]: m_maxmin_dis=0 to retain default min/max latching even though not used
	[3]: m_maxmin_dis_ths=0
	[2]: m_maxmin_rst=0
	[1-0]: m_rst_cnt=00 to enable magnetic reset each cycle

M_CTRL_REG3 (address 0x5D): write 0000 0000 = 0x00 to magnetometer control register 3
	[7]: m_raw=0: values stored in M_OFF_X/Y/Z offset registers are applied to magnetic sample data
	[6-4]: m_aslp_os=0: defines magetometer oversample ratio (OSR) in sleep mode (not used)
	[3]: m_ths_xyz_update=0: defines which reference values are updated when mag threshold event detection triggered (not used)
	[2-0]: reserved


FXAS21002C Gyro settings

CTRL_REG0 (address 0x0D): write 0001 1110 = 0x1E to gryo control register 0
	[7-6]: BW: bandwidth, cutoff freq of low pass filter = 4Hz at 12.5Hz ODR
	[5]: SPI interface mode
	[4-3]: SEL: high pass cutoff freq selection = 0.031Hz at 12.5Hz ODR
	[2]: HPF_EN: high pass filter enable = enabled
	[1-0]: FS: full scale range selection = 500 DPS

CTRL_REG1 (address 0x13): write 0001 1110 = 0x1E
	[7]: -
	[6]: software reset
	[5]: self-test enable
	[4-2]: Output Data Rate (ODR) selection = 111 (12.5Hz)
	[1]: standby/active mode = active
	[0]: standby/ready mode = 0



microSWIFT/utils/configuration.py
"""
Definition of the configuration class for the microSWIFT. The
configuration will read in a few variables from the
"""

from datetime import datetime, timedelta
import logging
import os
import warnings

from . import log

class Config:
    """
    Class object for configuration of the microSWIFT.
    """
    def __init__(self, config_fname: str):
        """
        Initialize the configuration of the microSWIFT with all
        settings. The settings include values that really should not
        be changed and are hardcoded into the class and parameters that
        can be regularly changed from the config.txt file.

        Parameters
        ----------
        config_fname : str
            File that contains simple parameters for configuration of
            the microSWIFT that can be easily changed.

        Returns
        -------
        config : class
            A class that contains all the parameters for the microSWIFT
            configuration including user changeable settings and
            hardcoded settings.
        """
        logger = logging.getLogger('microSWIFT')

        (duty_cycle_length,
         record_window_length,
         gps_sampling_frequency,
         imu_sampling_frequency) = self.read_config_file(config_fname)

        # Check read in parameters
        if record_window_length >= duty_cycle_length:
            raise ValueError('record_window_length must be strictly less than'
                             'the duty_cycle_length')
        else:
            pass

        if (duty_cycle_length - record_window_length) < 5:
            warnings.warn('send and process window SHOULD be longer than 5'
                          ' minutes')

        if gps_sampling_frequency >=1 and gps_sampling_frequency <= 4:
            pass
        else:
            raise ValueError('gps_sampling_frequency is out of range')

        if imu_sampling_frequency >=1 and imu_sampling_frequency <= 30:
            pass
        else:
            raise ValueError('imu_sampling_frequency is out of range')

        # Timing Parameters
        self.DUTY_CYCLES_PER_HOUR = int(60/duty_cycle_length)
        self.DUTY_CYCLE_LENGTH = timedelta(minutes=duty_cycle_length)
        self.RECORD_WINDOW_LENGTH = timedelta(minutes=record_window_length)
        self.START_TIME = self.get_start_time()
        self.END_RECORD_TIME = self.START_TIME + self.RECORD_WINDOW_LENGTH
        self.END_DUTY_CYCLE_TIME = self.START_TIME + self.DUTY_CYCLE_LENGTH

        # System Parameters
        self.ID = os.uname()[1]
        self.PAYLOAD_TYPE = 7
        self.SENSOR_TYPE = 52

        # GPS Parameters
        self.GPS_SAMPLING_FREQ = gps_sampling_frequency
        self.gpsGPIO = 21
        self.GPS_PORT = '/dev/ttyS0'
        self.START_BAUD = 9600
        self.BAUD = 115200

        # IMU Parameters
        self.IMU_SAMPLING_FREQ = imu_sampling_frequency
        self.IMU_GPIO = 20

        # Data Parameters
        self.WAVE_PROCESSING_TYPE = 'gps_waves'
        self.BAD_VALUE = 999
        self.NUM_COEF = 42

        logger.info(log.header(''))
        logger.info('Booted up. microSWIFT configuration: \n'
                    f'float ID: {self.ID}, payload type: {self.PAYLOAD_TYPE},'
                    f' sensor type: {self.SENSOR_TYPE},'
                    f'duty cycle length: {self.DUTY_CYCLE_LENGTH},'
                    f'record window length: {self.RECORD_WINDOW_LENGTH}')

        return

    def read_config_file(self, config_fname):
        """
        Read configuration parameters from the configuration file.

        Parameters
        ----------
        config_fname : str
            File that contains simple parameters for configuration of
            the microSWIFT that can be easily changed.

        Returns
        -------


        """
        try:
            config_file = open(config_fname, 'r')
        except:
            raise FileNotFoundError('cannot find config.txt file')

        lines = config_file.readlines()

        for line in lines:
            vals = line.split()
            if 'duty_cycle_length' in vals:
                try:
                    duty_cycle_length = int(vals[-1])
                except:
                    raise ValueError('duty_cycle_length must be an integer')

            elif 'record_window_length' in vals:
                try:
                    record_window_length = int(vals[-1])
                except:
                    raise ValueError('record_window_length must be an integer')

            elif 'gps_sampling_frequency' in vals:
                try:
                    gps_sampling_frequency = int(vals[-1])
                except:
                    raise ValueError('gps_sampling_frequency must be an'
                                     'integer')

            elif 'imu_sampling_frequency' in vals:
                try:
                    imu_sampling_frequency = int(vals[-1])
                except:
                    raise ValueError('imu_sampling_frequnecy must be an'
                                     'integer')
            else:
                continue

        return (duty_cycle_length,
                record_window_length,
                gps_sampling_frequency,
                imu_sampling_frequency)

    def get_start_time(self):
        """
        Find the start of the current duty cycle as a datetime.
        """
        current_hour = datetime.utcnow().replace(minute=0, second=0)
        for i in range(self.DUTY_CYCLES_PER_HOUR):
            possible_start_time = current_hour + i*self.DUTY_CYCLE_LENGTH
            if possible_start_time <= datetime.utcnow() \
                                    < possible_start_time + self.DUTY_CYCLE_LENGTH:
                start_time = possible_start_time
            else:
                continue

        return start_time

    def update_times(self):
        """
        Update timing parameters.
        """
        self.START_TIME = self.START_TIME + self.DUTY_CYCLE_LENGTH
        self.END_RECORD_TIME = self.START_TIME + self.RECORD_WINDOW_LENGTH
        self.END_DUTY_CYCLE_TIME = self.START_TIME + self.DUTY_CYCLE_LENGTH
